rewrite-core/src/main/java/org/openrewrite/RecipeScheduler.java

```diff
@@ -161,26 +164,58 @@ default RecipeRun scheduleRun(Recipe recipe,
     default <S extends SourceFile> List<S> scheduleVisit(RecipeRunStats runStats,
                                                          Stack<Recipe> recipeStack,
                                                          List<S> before,
+                                                         @Nullable List<Boolean> singleSourceApplicableTestResult,
                                                          ExecutionContext ctx,
                                                          Map<UUID, Stack<Recipe>> recipeThatAddedOrDeletedSourceFile) {
         runStats.calls.incrementAndGet();
         long startTime = System.nanoTime();
         Recipe recipe = recipeStack.peek();
+
         ctx.putCurrentRecipe(recipe);
         if (ctx instanceof WatchableExecutionContext) {
             ((WatchableExecutionContext) ctx).resetHasNewMessages();
         }
         try {
-            for (Recipe r : recipeStack) {
-                nextTest:
-                for (TreeVisitor<?, ExecutionContext> applicableTest : r.getApplicableTests()) {
-                    for (S s : before) {
-                        if (applicableTest.visit(s, ctx) != s) {
-                            continue nextTest;
+            if (!recipe.getApplicableTests().isEmpty() && !recipe.getSingleSourceApplicableTests().isEmpty()) {
+                throw new IllegalArgumentException("It's not allowed to have both `singleSource` and `anySource` since it's ambiguous. ");
+            }
```

Comment:
Ambiguity is resolved with a logical `AND`... so if there were both an any source and single source, both criteria would have to be met for the recipe to run